src/library_wasmfs.js

```diff
@@ -255,8 +255,21 @@ FS.createPreloadedFile = FS_createPreloadedFile;
     // TDOO: chown
     // TODO: lchown
     // TODO: fchown
-    // TODO: truncate
-    // TODO: ftruncate
+    truncate: (path, len) => {
+      if (len < 0) {
+        throw new FS.ErrnoError({{{ cDefs.EINVAL }}});
+      }
```

Comment:
If one calls `truncate("foo", -1)` I would probably expect it to try to resize of MAX_INT or something like that.. so maybe we don't need this check and we can just let the underlying call fail?

Reply:
That's what I was thinking, but I had one test that tried calling `truncate` with -10. I tried printing out the value the syscall receives, which was `4294867286`. But instead of setting the return value to -1 and setting `errno` accordingly, I think the syscall aborts. 

```
Aborted()
RuntimeError: Aborted(). Build with -sASSERTIONS for more info.
    at abort (/home/ubuntu/Documents/emsdk/emscripten/out/test/test_fs_js_api.js:1:5674)
    at _abort (/home/ubuntu/Documents/emsdk/emscripten/out/test/test_fs_js_api.js:1:14749)
    at <anonymous>:wasm-function[187]:0xd2a9
    at <anonymous>:wasm-function[197]:0xe13e
    at <anonymous>:wasm-function[172]:0xba9a
    at <anonymous>:wasm-function[178]:0xce81
    at _wasmfs_truncate (<anonymous>:wasm-function[261]:0x13101)
```

This is the error. It doesn't seem to match the pattern of how we've been returning JS API errors before (Which throws an error from JS with an error name and code). Is this close to what you were saying by letting the call fail?

Reply:
I agree we don't want to syscall aborting like that.   If you build with `-g` or `--profiling-funcs` you should get functions names for those wasm functions in the backtrace BTW.

Reply:
This is strange because `doTruncate` does return -EINVAL when its input is negative. It would be good to get to the bottom of this!

Reply:
I noticed that `doTruncate` takes in `off_t size`, but `truncate64` and `ftruncate64` take in size as a `uint64_t`, so is it possible that the size is always positive in `doTruncate`?

Reply:
Is it possible that this issue is caused because there is no method to send i64 from JS to Wasm, like we discussed on the Chat space? I've tried changing `truncate64` to take in a signed `int64_t`, but it hasn't fixed the issue yet.

Reply:
I think this is needed for the Node backend, as it silently accepts negative offsets rather than failing with `EINVAL`, see: https://github.com/nodejs/node/issues/35632.

Though, it looks like `setSize()` is currently not implemented for the Node backend :
https://github.com/emscripten-core/emscripten/blob/48b0645d3075ab8003745d2c13c60f514fa4063a/system/lib/wasmfs/backends/node_backend.cpp#L120-L122
(i.e. this PR won't work for `-sNODERAWFS` + `-sWASMFS`)

Reply:
If `setSize()` (and `insertMove()` due to `FS.rename()`) is supported in WasmFS' Node backend (which is probably better as a follow-up), you could do:
```diff
--- a/test/test_core.py
+++ b/test/test_core.py
@@ -6017,6 +6017,7 @@ Module = {
   def test_fs_writeFile(self):
     self.do_run_in_out_file_test('fs/test_writeFile.cpp')
 
+  @also_with_noderawfs
   def test_fs_js_api(self):
     self.set_setting("FORCE_FILESYSTEM")
     self.do_runf(test_file('fs/test_fs_js_api.c'), 'success')
```

After that, you can verify with:
```console
$ ./test/runner wasmfs.test_fs_js_api* core2.test_fs_js_api*
```

Whether the test passes on both memory and Node backend, either with or without WasmFS.